lucille-core/src/test/java/com/kmwllc/lucille/util/PreemptiveAuthInterceptorTest.java

```diff
@@ -0,0 +1,76 @@
+package com.kmwllc.lucille.util;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.when;
+import static org.mockito.Mockito.any;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import org.apache.http.Consts;
+import org.apache.http.HttpException;
+import org.apache.http.HttpHost;
+import org.apache.http.HttpRequest;
+import org.apache.http.auth.AuthScope;
+import org.apache.http.auth.AuthState;
+import org.apache.http.auth.Credentials;
+import org.apache.http.client.CredentialsProvider;
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.impl.auth.BasicScheme;
+import org.apache.http.protocol.HttpContext;
+import org.junit.Test;
+import org.mockito.ArgumentCaptor;
+
+public class PreemptiveAuthInterceptorTest {
+
+  @Test 
+  public void testProcess() throws Exception {
+    AuthState nullState = mock(AuthState.class);
+    AuthState  state = mock(AuthState.class);
+    
+    when(state.getAuthScheme()).thenReturn(new BasicScheme());
+    when(nullState.getAuthScheme()).thenReturn(null);
+
+    Credentials credentials = mock(Credentials.class);
+    CredentialsProvider credsProvider = mock(CredentialsProvider.class);
+    when(credsProvider.getCredentials(new AuthScope("foo", 1))).thenReturn(credentials);
+    when(credsProvider.getCredentials(new AuthScope("bar", 1))).thenReturn(null);
+
+    HttpHost targetHostFoo = new HttpHost("foo", 1);
+    HttpHost targetHostBar = new HttpHost("bar", 1);
+
+    HttpContext nullContext = mock(HttpContext.class);
```

Comment:
for readibility, could you create each mock and finishing setting it up, one at a time? This would mean instantiating nullContext, then doing all the "when" statements that affect it, before moving on to the next mock for context, etc. This way, someone could understand the setup of each mock without having to skip around to multiple places in the code down below where setup for different mocks is interleaved.